### DISCUSSION TITLE
How Do I Retreive Meta Message Values?

### DISCUSSION THREAD

**Discussion by JeffRocchio** [mido/mido#575] [Q&A] [answered]
I am trying to read, from a .mid file, a few key meta message values, such as *Key Signature*, *Time Signature*, etc. I can detect them using statements like:
`if msg.type == 'time_signature':
                print('--- TIME SIGNATURE MESSAGE DETECTED ---)`

But I am unable to work out how to then retreive the values, e.g., the numerator= and denominator= parts.

I have hunted around for about an hour now to find a code example for doing this, but all I can find are examples of *writing* a MetaMessage, none for reading one.

**NOTE**: I am more of a C++ guy, some Python experience, but not all that much. So my lack of understanding may be down to some combo of unfamiliarity with Python and, of course, none with mido (as yet).

Any help appreciated!

**Comment by rdoursenaud** [comment, marked as answer]
Hi, and thank you for using mido!

All the attributes are [docummented](https://mido.readthedocs.io/en/stable/meta_message_types.html).
For the `time_signature` [object](https://mido.readthedocs.io/en/stable/meta_message_types.html#time-signature-0x58) you'll find the attributes:
- `numerator`
- `denominator`
- `clocks_per_click`
- `notated_32nd_notes_per_beat`

You simply have to read them from the object:
```py
if msg.type == 'time_signature':
    print(f"Numerator: {msg.numerator}")
    print(f"Denominator: {msg.denominator}")
    ...
```

Let me know if you need further clarification.

**Comment by JeffRocchio** [comment]
That's too easy! But, I just tried it and it works. Thanks for the quick response Raphaël.

**Reply by rdoursenaud** [reply]
That's the point of mido: provide high level object representations of MIDI one can easily interact with.
I'm glad you enjoy it :)
